Create a new house and add to a town

TuDucThinh opened the pull request:
- Description of add house feature:
   + To add a house , you need conditional this town still exist.
   + To add a house you need: house number and town ID
- Requests in the same town cannot have the same house number.
- The house number must have at least 2 characters, if the house number is the numbers from 1 to 9, add a 0 in front

TuDucThinh committed:
add House class and HomeFunction

TuDucThinh committed:
Delete Main.java

TuDucThinh committed:
Update isNotValidStreetName

fix 48 and 50 . lines

ntananh committed:
add class diagram for entity

TuDucThinh committed:
add Management class and add extra function House

TuDucThinh committed:
Add HouseManagement class and add extra function House

TuDucThinh committed:
Merge branch 'FEAT-TDT-008--create-a-new-house-and-add-to-a-town' of https://github.com/n-tananh/resitrack into FEAT-TDT-008--create-a-new-house-and-add-to-a-town

TuDucThinh committed:
Merge branch 'main' into FEAT-TDT-008--create-a-new-house-and-add-to-a-town

# Conflicts:
#	src/main/java/org/resitrack/entity/House.java
#	src/main/java/org/resitrack/entity/Person.java
#	src/main/java/org/resitrack/entity/Town.java

TuDucThinh committed:
add sample house functions

TuDucThinh committed:
edit the HouseManagement class, edit the House class. add the function of adding multiple houses at once.

TuDucThinh committed:
add java doc to HouseController class and MenuUtil class

TuDucThinh committed:
delete the diagram entity class image

TuDucThinh committed:
edit function validateStreetName

TuDucThinh committed:
edit java doc and edit function add house to town

TuDucThinh committed:
- in HouseController class:
      + Add and edit java doc.
      + Edit function createNewHouse().
      + Delete getStreetName, validateStreet name
- in House class:
      + Delete getter Street Name,

TuDucThinh committed:
Delete import java.util.ArrayList and delete function printAllHouseInformation.
Edit java doc

ntananh commented on src/main/java/org/resitrack/Main.java:
```diff
@@ -72,4 +75,25 @@ public static void managePerson(PersonController personController, Scanner scann
+            choice = scanner.nextLine();
+
+            switch (choice) {
+                case "1":
```
use enum

TuDucThinh committed:
- Edit Main class:
   + Edit manageHouse
   + Delete import java.util.Arrays
- Edit HouseController class:
   + Edit java doc
   + Delete getHouseInfo
   + Delete import org.resitrack.entity.Town
- Edit MenuUtil class:
   +  Add     public final static String ADD_HOUSE = "1";
   + Delete option 2 in houseMenu.

ntananh commented on src/main/java/org/resitrack/Main.java:
```diff
@@ -34,15 +34,17 @@ public static void main(String[] args) {
-                    System.out.println("Manage house");
+                    houseController.setHouses(houses);
+                    townController.setTowns(towns);
+                    houseController.setTownController(townController);
```
I think we should put these line in to the `manageHouse` function

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+import java.util.Scanner;
+
+/**
+ * Create a House Controller.
```
For the class's Javadoc, we should write it to explain the purpose of this class
``` suggestion
 * House Controller used for managing houses, allow user to manipulate with house data.
```

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+     *
+     * @param houseNumber pass the house number as String
+     * @param townId      pass ID town in String
+     * @return {@link House} New house created
```
```suggestion
     * @return {@link House} new created house
```

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+    }
+
+    /**
+     * This function is used to check the error of the house number entered by the user and give a message
```
```suggestion
     *  Validation function is used to check the error of the house number entered by the user and give a message
```

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+     *
+     * @param houseNumber new house number
+     * @param townId      town id to check if given house number exist or not
+     * @return Errors when user enters house number
```
```suggestion
     * @return Error message when user enters an invalid house number
```

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+            System.out.println("\nAdded " + numberOfHouse + " house. The latest home address is: " + houseNumber + " " + townName);
+            System.out.println();
+            System.out.print("Enter 0 to exit, Enter 1 to continue: ");
+            choice = scanner.nextInt();
```
What happens when I input a string value in this line?  Is it still valid for the value you use in while loop for check

TuDucThinh replied:
if you enter a string here, it will stop the program immediately and give an error message

ntananh replied:
It will face with an exception, that mean the app crashed. Do you think we expect when user input wrong data and the app stop working.

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+    public void createNewHouse() {
+
+        String townId = getTownId();
+        String townName = townController.getTownById(townId).getName();
```
As we discussed before, we should create a function that returns a valid Town. A Town contains id and name so we don't need to access into town list twice time to get information of same object

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,178 @@
+            System.out.print("Enter 0 to exit, Enter 1 to continue: ");
+            choice = scanner.nextInt();
+            scanner.nextLine();
+            if (choice == 1) {
+                numberOfHouse++;
+            }
```
We can remove this check by make choice is String type. 

```
        String choice;
        do {
            System.out.println("Adding a new house");
            String houseNumber = getHouseNumber(townId);
            this.houses.add(createHouse(houseNumber, townId));

            System.out.println("\nAdded " + numberOfHouse + " house. The latest home address is: " + houseNumber + " " + townName);
            System.out.print(">Enter 0 to exit, or press any key to continue!");
            choice = scanner.nextLine();
            numberOfHouse++;

        } while (!choice.equalsIgnoreCase("0"));
        
  ```

TuDucThinh replied:
ok i will replace it now

ntananh reviewed (commented):
Hi Thinh, I've finished second review. Feel free to take a look 😄

TuDucThinh committed:
- Main class:
   + Edit manageHouse
- HouseController class:
   + Edit java doc
   + Edit creatNewHouse

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,184 @@
+        return townId;
+    }
+
+    private String idName(Town town) {
+
+        return town.getName();
+    }
+
```
what is the purpose of this function ?

TuDucThinh replied:
Sorry. I forgot to delete it

ntananh commented on src/main/java/org/resitrack/Main.java:
```diff
@@ -72,4 +69,31 @@ public static void managePerson(PersonController personController, Scanner scann
+        List<Town> towns = new ArrayList<>();
+
+        HouseController houseController = new HouseController();
+        TownController townController = new TownController();
```
we created the TownController object at line 23, why we need to create it again ?

TuDucThinh replied:
I spent all afternoon fixing this place but it didn't work. because if I don't initialize it it won't let me access it through townController and houseController

TuDucThinh replied:
Do you have any ideas for this? please suggest me

TuDucThinh replied:
I have successfully fixed it here

ntananh commented on src/main/java/org/resitrack/Main.java:
```diff
@@ -72,4 +69,31 @@ public static void managePerson(PersonController personController, Scanner scann
+    public static void manageHouse(Scanner scanner) {
+        String choice;
+        List<House> houses = new ArrayList<>();
+        List<Town> towns = new ArrayList<>();
```
why we need create a list of towns object again here ?

TuDucThinh replied:
I fixed it

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,184 @@
+        String townId = town.getId();
+        String townName = town.getName();
+
+        int numberOfHouse = 1;
```
i think we should check town before init this variable here

TuDucThinh replied:
i fixed it

ntananh commented on src/main/java/org/resitrack/controller/HouseController.java:
```diff
@@ -1,4 +1,184 @@
+            choice = scanner.nextLine();
+            numberOfHouse++;
+
+        } while (!choice.equalsIgnoreCase("0"));
```
I think we should use enum here.

TuDucThinh replied:
I fixed it

ntananh reviewed (commented):
Done second review, feel free to take a look

TuDucThinh committed:
- Edit Main
    + Add java doc
    + Edit main and manageHouse
- Edit HouseController
    +  Add java doc
    + Edit function createNewHouse
    + Delete function idName

TuDucThinh commented:
I have successfully fixed the errors you mentioned above

n-tananh commented:
All fixes is oke, ready to merge

ntananh reviewed (approved):
Finished review
